### ISSUE TITLE
Solution vectors are not restored when solve fails in sub-apps while doing Picard iterations

### ISSUE THREAD

**Issue by andrsd** [idaholab/moose#5827]
When running BISON (master) and RELAP-7 (sub-app), it can happen that RELAP-7 gets into a non-physical region which results in a NaN in the solution vector. MOOSE correctly cuts the time step in a half and tries to resolve, however, the initial vector on RELAP-7 side still has a NaN, so the solve aborts immediately, time steps gets cut again and so on until we reach dt_min.

**Comment by friedmud** [comment]
Are you doing subcycling or not?

**Comment by andrsd** [comment]
No sub-cycling.

**Comment by friedmud** [comment]
All of the calls in `TransientMultiApp` to `incrementStepOrReject()` ultimately call `TimeStepper::rejectStep()` which will call `FEProblem::restoreSolutions()` ( https://github.com/idaholab/moose/blob/devel/framework/src/timesteppers/TimeStepper.C#L207 ) which should do the right thing.

Here are the `incrementStepOrReject()` calls in `TransientMultiApp`:
https://github.com/idaholab/moose/blob/devel/framework/src/multiapps/TransientMultiApp.C#L208
https://github.com/idaholab/moose/blob/devel/framework/src/multiapps/TransientMultiApp.C#L317

Can you see why you're not hitting them?

Can you paste your `[MultiApp]` block so I can see what options you're using?

Trying to force restoration from the outside (master) is NOT a good idea.  You have to remember that there might be sub-apps underneath this one that are Picarding within this sub-app!  This has to be handled carefully... and all the right code is already in the `Transient` Executioner (like in `incrementStepOrReject()`).

**Comment by friedmud** [comment]
Are you allowing "catch up steps"?

If not... then the master app should be restoring all MultiApps from `Backup` when the solve of the current step fails... so there should be no need for a solution restore.

**Comment by andrsd** [comment]
Here is the block
```
  [./relap]
    type = TransientMultiApp
    app_type = RELAP7App
    execute_on = timestep_end
    max_procs_per_app = 1
    sub_cycling = false
    output_sub_cycles = true
    detect_steady_state = true
    positions = '0 0 0'
    input_files = relap.i
  [../]

```

> Can you see why you're not hitting them?

The first reference to `incrementStepOrReject` is irrelevant to my case, since it is in the sub-cycling case.

The second one won't be called at all because my `picard_max_its = 3` which sets the `auto_advance` flag to false. See

https://github.com/idaholab/moose/blob/devel/framework/src/executioners/Transient.C#L390

and

https://github.com/idaholab/moose/blob/devel/framework/src/executioners/Transient.C#L443

**Comment by friedmud** [comment]
In your case... when the sub-app fails to solve it should fail out in the master app and you should end up here in the master app's Transient Executioner:

https://github.com/idaholab/moose/blob/devel/framework/src/executioners/Transient.C#L307

Those `restoreMultiApp()` calls should wipe any memory of what what RELAP-7 just tried to do and completely reset it back to the state it was in before it tried to solve this step... including resetting the solution vector.

Can you verify that that's the sequence that's happening?  Can you dig out and print out the RELAP-7 solution vector before and after those `restoreMultiApp()` calls?

**Comment by andrsd** [comment]
I think that was not happening. When RELAP-7 failed, even though the `NonlinearSystem` reported non-convergence, the `_multiapps_converged` variable was set to true. That's why I had to add the else-statement into `TransientMultiApp`. But, when I ran with that I still saw the code to cut the time steps. When I added that `restoreMultiAppsSolutions` it actually finished.

I agree with you that my fix is not right. I will undo `restoreMultiAppsSolutions` and look at the vectors.

I also think that the `auto_advance` flag prevents proper call to `incrementStepOrReject()` in the `TransientMultiApp`.

**Comment by friedmud** [comment]
Then the issue is that `_multiapps_converged` wasn't right... can you look into that?

I may be able to look into it later tonight if you have a specific example that shows the issue.

**Comment by andrsd** [comment]
> can you look into that?

Yes, that's my plan. I'll let you know when I have something.

**Comment by friedmud** [comment]
What's the status here?  I'm tied up today until about 12:30PM your time... but I can help this afternoon.

**Comment by andrsd** [comment]
I need to do something for RELAP-7 (user support) today. Then, I'd work on it. What I was thinking to do was to replicate this in a simplified `moose_test` setup, where the sub-app would be generating NaNs based on some conditions (specified in and input file). The simulation that was exhibiting this is in mammoth and has to run for quite a while (~ hours) before I hit the bug.

**Comment by friedmud** [comment]
Yeah - if you can make up a small example that does it... I can definitely fix it this afternoon.

**Comment by andrsd** [comment]
Initial set of input files. It does not have the behavior that I originally described, but you can see that when the sub app generates a NaN and fails solving, the multi system reports:
```
Successfully Solved MultiApp sub.
```
That's all I have for right now.

The original behavior seems to be difficult to replicate, because the NaNs are coming from the EOS package so it is not that easy as `put a NaN in the residual vector`.

**Comment by friedmud** [comment]
Thanks! That should be good enough... I'll take a look...

**Comment by andrsd** [comment]
So I played with the code a bit. I can fix the problem that `_multiapps_converged` was true, even if the solve failed by adding that else-branch in `TransientMultiApp` (see my PR for diff). I see that the code cuts the time step in a half for the next iteration, but I do not see `rejectStep()` being called. And when printing the solution vector, the initial values are different than the ones in the previous time step.

Not sure if that helps you in any way...

**Comment by friedmud** [comment]
It does... I'm digging in now.  I'll let you know soon.

**Comment by friedmud** [comment]
I have this figured out.  Working up a PR now with a test...

**Comment by friedmud** [comment]
Ok - fix is in.  This ended up being a _few_ issues all piled up:

1.  We weren't correctly returning that the sub-app failed in the case where we weren't doing sub_cycling or catch-up steps (most likely because I never use this! ;-)

2.  Even if we were... we weren't taking the correct action in Transient Executioner.  The issue is that we were in the Picard loop... and we didn't exit if multiapps didn't converge.  Ooops.

3.  EVEN IF those two things were fixed... the sub-app still wasn't getting restored properly.  The reason why is because of #5554 (selective restoration).  This configuration (no sub_cycling or catch_up_steps) is a candidate for selective restoration (where we don't restore MultiApps that meet the criteria at the beginning of each Picard solve because it helps them solve faster).  This was suppressing restoration.

To fix (3) I added a `force` option on `FEProblem::restoreMultiApps()`... so that in the case of solve failure we can force the state of a sub-app to get reset.

All three of these things together give the proper behavior.  I even went down and printed out solution vectors to make sure.

Merge the PR and give it a whirl and let me know if it works....
